source/common/grpc/google_async_client_impl.cc

```diff
@@ -0,0 +1,371 @@
+  // TODO(htuch): add support for SSL, OAuth2, GCP, etc. credentials.
+  std::shared_ptr<grpc::ChannelCredentials> creds = grpc::InsecureChannelCredentials();
+  // We rebuild the channel each time we construct the channel. It appears that the gRPC library is
+  // smart enough to do connection pooling and reuse with identical channel args, so this should
+  // have comparable overhead to what we are doing in Grpc::AsyncClientImpl, i.e. no expensive
+  // new connection implied.
+  std::shared_ptr<grpc::Channel> channel = CreateChannel(config.target_uri(), creds);
+  stub_ = std::make_unique<grpc::GenericStub>(channel);
+  // Initialize client stats.
+  stats_.streams_total_ = &scope_.counter("streams_total");
+  for (uint32_t i = 0; i <= Status::GrpcStatus::MaximumValid; ++i) {
+    stats_.streams_closed_[i] = &scope_.counter(fmt::format("streams_closed_{}", i));
+  }
+}
+
+GoogleAsyncClientImpl::~GoogleAsyncClientImpl() {
+  while (!active_streams_.empty()) {
+    active_streams_.front()->resetStream();
+  }
+}
+
+AsyncRequest* GoogleAsyncClientImpl::send(const Protobuf::MethodDescriptor& service_method,
+                                          const Protobuf::Message& request,
+                                          AsyncRequestCallbacks& callbacks,
+                                          Tracing::Span& parent_span,
+                                          const Optional<std::chrono::milliseconds>& timeout) {
+  auto* const async_request =
+      new GoogleAsyncRequestImpl(*this, service_method, request, callbacks, parent_span, timeout);
+  std::unique_ptr<GoogleAsyncStreamImpl> grpc_stream{async_request};
+
+  grpc_stream->initialize(true);
+  if (grpc_stream->call_failed()) {
+    return nullptr;
+  }
+
+  grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+  return async_request;
+}
+
+AsyncStream* GoogleAsyncClientImpl::start(const Protobuf::MethodDescriptor& service_method,
+                                          AsyncStreamCallbacks& callbacks) {
+  const Optional<std::chrono::milliseconds> no_timeout;
+  auto grpc_stream =
+      std::make_unique<GoogleAsyncStreamImpl>(*this, service_method, callbacks, no_timeout);
+
+  grpc_stream->initialize(false);
+  if (grpc_stream->call_failed()) {
+    return nullptr;
+  }
+
+  grpc_stream->moveIntoList(std::move(grpc_stream), active_streams_);
+  return active_streams_.front().get();
+}
+
+GoogleAsyncStreamImpl::GoogleAsyncStreamImpl(GoogleAsyncClientImpl& parent,
+                                             const Protobuf::MethodDescriptor& service_method,
+                                             AsyncStreamCallbacks& callbacks,
+                                             const Optional<std::chrono::milliseconds>& timeout)
+    : parent_(parent), service_method_(service_method), callbacks_(callbacks), timeout_(timeout),
+      completion_thread_(new Thread::Thread([this] { completionThread(); })) {}
+
```

Comment:
This seems technically sound, but let me say that it runs in the face of "best practices." Having a separate completion thread per stream is a serious underutilization of the async API, IMO. In fact, I would say that the lower thread usage of the async API is really its biggest motivation compared to the sync API. Can you add a TODO here itself at the least to mention something about future plans about this? I presume that this is a transitional implementation.

Reply:
Yep, agree, this is transitional, will have a PR up shortly to do the TLS CQ.